csharp/src/Ice/AsyncSemaphore.cs

```diff
@@ -46,31 +63,42 @@ internal async ValueTask WaitAsync(CancellationToken cancel = default)
             }
             catch (Exception ex)
             {
-                taskCompletionSource.TrySetException(ex);
+                lock (_mutex)
+                {
+                    if (_queue.Contains(taskCompletionSource))
+                    {
+                        taskCompletionSource.SetException(ex);
+                    }
+                }
                 throw;
             }
         }
 
         internal void Release()
         {
-            ManualResetValueTaskCompletionSource<bool> taskCompletionSource;
-            do
+            lock (_mutex)
             {
-                lock (_mutex)
+                if (_currentCount == _maxCount)
                 {
-                    if (_queue.Count == 0)
+                    throw new SemaphoreFullException($"semaphore maximum count of {_maxCount} already reached");
+                }
+
+                while (_queue.Count > 0)
+                {
+                    try
```

Comment:
Right, SetResult will throw if the task is already completed. This can occur if WaitAsync is canceled. For example, if an invocation is waiting on the semaphore and the invocation timeout kicks in.

Another option would be for the WaitAsync implementation to remove the queued task from the queue but this would require to use a collection that supports a Remove operation.